pkg/nodeagent/files/filesystem.go

```diff
@@ -0,0 +1,453 @@
+	if err != nil {
+		return nil, fmt.Errorf(getStateErrorMessage, name, err)
+	}
+
+	file, err := n.fs.OpenFile(name, flag, perm)
+	if err != nil {
+		return file, err
+	}
+
+	if err := n.saveState(name, operation); err != nil {
+		return nil, fmt.Errorf(saveStateErrorMessage, name, err)
+	}
+
+	return file, nil
+}
+
+// Remove removes a file identified by name, returning an error, if any
+// happens.
+func (n *nodeAgentFileSystem) Remove(name string) error {
+	if err := n.fs.Remove(name); err != nil {
+		return err
+	}
+
+	if err := n.deleteState(name); err != nil {
+		return fmt.Errorf(saveStateErrorMessage, name, err)
+	}
+
+	return nil
+}
+
+// RemoveAll removes a directory path and any children it contains. It
+// does not fail if the path does not exist (return nil).
+func (n *nodeAgentFileSystem) RemoveAll(path string) error {
+	if err := n.fs.RemoveAll(path); err != nil {
+		return err
+	}
+
+	if err := n.deleteState(path); err != nil {
+		return fmt.Errorf(saveStateErrorMessage, path, err)
+	}
+
+	for nodeAgentFile := range n.nodeAgentFsOperations.FileSystemOperations {
+		if !strings.HasPrefix(nodeAgentFile, path) {
+			continue
+		}
+		if err := n.deleteState(nodeAgentFile); err != nil {
+			return fmt.Errorf(saveStateErrorMessage, nodeAgentFile, err)
+		}
+	}
+
+	return nil
+}
+
+// Rename renames a file.
+func (n *nodeAgentFileSystem) Rename(oldname, newname string) error {
+	operation, err := n.getState(newname)
+	if err != nil {
+		return fmt.Errorf(getStateErrorMessage, newname, err)
+	}
+
+	if err := n.fs.Rename(oldname, newname); err != nil {
+		return err
+	}
+
+	if err := n.saveState(newname, operation); err != nil {
+		return fmt.Errorf(saveStateErrorMessage, newname, err)
+	}
+
+	if err := n.deleteState(oldname); err != nil {
+		return fmt.Errorf(saveStateErrorMessage, oldname, err)
+	}
+
+	return nil
+}
+
+// Stat returns a FileInfo describing the named file, or an error, if any
+// happens.
+func (n *nodeAgentFileSystem) Stat(name string) (os.FileInfo, error) {
+	return n.fs.Stat(name)
+}
+
+// Name returns the name of this FileSystem
+func (n *nodeAgentFileSystem) Name() string {
+	return "NodeAgentFileSystem"
+}
+
+// Chmod changes the mode of the named file to mode.
+func (n *nodeAgentFileSystem) Chmod(name string, mode os.FileMode) error {
+	return n.fs.Chmod(name, mode)
+}
+
+// Chown changes the uid and gid of the named file.
+func (n *nodeAgentFileSystem) Chown(name string, uid, gid int) error {
+	return n.fs.Chown(name, uid, gid)
+}
+
+// Chtimes changes the access and modification times of the named file
+func (n *nodeAgentFileSystem) Chtimes(name string, atime time.Time, mtime time.Time) error {
+	return n.fs.Chtimes(name, atime, mtime)
+}
+
+func (n *nodeAgentFileSystem) getState(path string) (*FileSystemOperation, error) {
+	if operation, ok := n.nodeAgentFsOperations.FileSystemOperations[path]; operation == OperationDeleted {
+		return ptr.To(OperationDeleted), nil
+	} else if ok {
+		return nil, nil
```

Comment:
nit: I spent a couple of minutes understanding why you return `nil` is an opertion exists in the map - Maybe the function name isn't the best, or we can add a comment explaining the behavior here. 
Because here returning `nil` means "I have already determined the state of this file", which is a surprising result of a function `getState()`